### PR TITLE
Fix Profiler + Minor improvements

### PR DESCRIPTION
**Pull request by JanKleine**
There is currently a bug in the profiler, that the first run recorded includes the runtime of all the warmup runs, making one ugly outlier. This is because `times[0]` is set before the warmups and is never properly reset.

This fixes that and makes the surrounding code a bit cleaner (e.g. using `np.diff` instead of list comprehension).

I also added some options that make the profiler more usable when it's called inside of other `tqdm` progress bars (allows setting `tqdm`'s `leave=False` and stop it from printing results).

List of changes:
- Fix: timing error
- Add: `tqdm_leave` parameter
- Add: `print_results` parameter
- Cleanup: Only print `Profiling...` if `tqdm` is not used
- Cleanup: use `warnings.warn()` instead of `print("WARNING: ...")`
- Cleanup: remove unused variable for return value (didn't seem to be used)
- Cleanup: make diff calculation more concise

### CODE REVIEW

File: dace/builtin_hooks.py [hunk 1]
```diff
@@ -35,6 +40,10 @@ def profile(repetitions: int = 100, warmup: int = 0):
 
     :param repetitions: The number of times to run each DaCe program.
     :param warmup: Number of additional repetitions to run the program without measuring time.
+    :param tqdm_leave: Sets the leave parameter of the tqdm progress bar (usefull
```
Comment (tbennun) [inline dace/builtin_hooks.py:43]:
```suggestion
    :param tqdm_leave: Sets the ``leave`` parameter of the ``tqdm`` progress bar (useful 
```

File: dace/builtin_hooks.py [hunk 2]
```diff
@@ -35,6 +40,10 @@ def profile(repetitions: int = 100, warmup: int = 0):
 
     :param repetitions: The number of times to run each DaCe program.
     :param warmup: Number of additional repetitions to run the program without measuring time.
+    :param tqdm_leave: Sets the leave parameter of the tqdm progress bar (usefull 
+        for nested progress bars). Ignored if tqdm progress bar is not used.
+    :param print_results: Whether or not to print the average execution time after
```
Comment (tbennun) [inline dace/builtin_hooks.py:45]:
```suggestion
    :param print_results: Whether or not to print the median execution time after 
```

File: dace/builtin_hooks.py [hunk 3]
```diff
@@ -35,6 +40,10 @@ def profile(repetitions: int = 100, warmup: int = 0):
 
     :param repetitions: The number of times to run each DaCe program.
     :param warmup: Number of additional repetitions to run the program without measuring time.
+    :param tqdm_leave: Sets the leave parameter of the tqdm progress bar (usefull 
+        for nested progress bars). Ignored if tqdm progress bar is not used.
+    :param print_results: Whether or not to print the average execution time after 
+        all repetitions (usefull to disable when using nested progress bars).
```
Comment (tbennun) [inline dace/builtin_hooks.py:46]:
```suggestion
        all repetitions.
```

File: dace/frontend/operations.py [hunk 4]
```diff
@@ -47,34 +57,45 @@ def __init__(self, repetitions: int = 0, warmup: int = 0) -> None:
...
-                      'dependency tqdm...\n\tTo see a live progress bar please install '
-                      'tqdm (`pip install tqdm`)\n\tTo disable this feature (and '
-                      'this warning) set `profiling_status` to false in the dace '
-                      'config (~/.dace.conf).')
+                warnings.warn(
+                    "Cannot show profiling progress, missing optional dependency "
+                    "tqdm...\n\tTo see a live progress bar please install tqdm "
+                    "(`pip install tqdm`)\n\tTo disable this feature (and this "
+                    "warning) set `profiling_status` to false in the dace config "
+                    "(~/.dace.conf)."
```
Comment (tbennun) [inline dace/frontend/operations.py:80]: why are the quotes double? see contributing guidelines.

File: dace/frontend/operations.py [hunk 5]
```diff
@@ -47,34 +57,45 @@ def __init__(self, repetitions: int = 0, warmup: int = 0) -> None:
...
+        # zeros to overwrite start time, followed by indices for each repetition
+        iterator = chain(repeat(0, self.warmup), range(1, self.repetitions + 1))
 
-        times = [start] * (self.repetitions + 1)
-        ret = None
-        print('\nProfiling...')
-
-        iterator = range(self.warmup + self.repetitions)
-        if Config.get_bool('profiling_status'):
+        if Config.get_bool("profiling_status"):
```
Comment (tbennun) [inline dace/frontend/operations.py:63]: double quotes, see contributing guidelines and apply everywhere

File: dace/frontend/operations.py [hunk 6]
```diff
@@ -47,34 +57,45 @@ def __init__(self, repetitions: int = 0, warmup: int = 0) -> None:
...
             try:
                 from tqdm import tqdm
-                iterator = tqdm(iterator, desc="Profiling", file=sys.stdout)
+
+                iterator = tqdm(
+                    iterator,
+                    desc="Profiling",
```
Comment (tbennun) [inline dace/frontend/operations.py:69]: double quotes

File: dace/frontend/operations.py [hunk 7]
```diff
@@ -105,7 +127,7 @@ def __call__(self, compiled_sdfg: 'CompiledSDFG', args: Tuple[Any, ...]):
         # Restore state after skipping contents
         compiled_sdfg.do_not_execute = old_dne
 
-        return ret
+        return None
```
Comment (tbennun) [inline dace/frontend/operations.py:130]: Why this change?

**Review by tbennun — CHANGES_REQUESTED** [review]

File: dace/frontend/operations.py (hunk 4 above)
Comment (phschaad) [reply dace/frontend/operations.py:80]: I don't think we have it in our contributing guidelines, do we? We just mention that we follow Google's style, and that just states to be consistent; which [we're not](https://github.com/spcl/dace/blob/master/dace/sdfg/sdfg.py#L183) 🙃

Comment (phschaad) [reply dace/frontend/operations.py:80]: Unless I'm missing it, we should probably add it

Comment (tbennun) [reply dace/frontend/operations.py:80]: oh dear, we should. thx!

File: dace/frontend/operations.py (hunk 7 above)
Comment (JanKleine) [reply dace/frontend/operations.py:130]: > - Cleanup: remove unused variable for return value (didn't seem to be used)

The variable didn't seem serve any purpose. So basically: "while I'm already here, might as well..."

File: dace/builtin_hooks.py [hunk 8]
```diff
@@ -40,10 +40,10 @@ def profile(
 
     :param repetitions: The number of times to run each DaCe program.
     :param warmup: Number of additional repetitions to run the program without measuring time.
-    :param tqdm_leave: Sets the leave parameter of the tqdm progress bar (usefull 
+    :param tqdm_leave: Sets the ``leave`` parameter of the ``tqdm`` progress bar (usefull
```
Comment (tbennun) [inline dace/builtin_hooks.py:43]:
```suggestion
    :param tqdm_leave: Sets the ``leave`` parameter of the ``tqdm`` progress bar (useful
```

**Review by tbennun — CHANGES_REQUESTED** [review]

**Review by tbennun — APPROVED** [review]
